### PR TITLE
Respect `transient` config for dbt Python models

### LINKED ISSUE #776: [ADAP-901] [Bug] Not possible to set `transient` config on Python models
### Is this a new bug in dbt-snowflake?

- [X] I believe this is a new bug in dbt-snowflake
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

1. It doesn't seem like it's possible to set the `transient` [config](https://docs.getdbt.com/reference/resource-configs/snowflake-configs#transient-tables) on Python models.
2. It's also defaulting to non-transient.

### Expected Behavior

`transient` config should be respected by the Python model.

### Steps To Reproduce

1. Project setup:

```yml
# dbt_project.yml
name: my_dbt_project
profile: all
config-version: 2
version: 1.0

models:
  my_dbt_project:
    +materialized: table
    # +transient: true    # This doesn't work for snek.py.
```

```python
# models/snek.py
import pandas as pd
def model(dbt, session):
    dbt.config(transient=True)
    # dbt.config(transient="true")    # Neither does using a string in a model level config.
    return pd.DataFrame({"id": [1]})
```

2. Run:

```sh
$ dbt --debug run -s snek

22:03:38  Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'start', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10a0b9820>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10910dbb0>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10ccf6130>]}
22:03:38  Running with dbt=1.6.2
22:03:38  running dbt with arguments {'printer_width': '80', 'indirect_selection': 'eager', 'log_cache_events': 'False', 'write_json': 'True', 'partial_parse': 'False', 'cache_selected_only': 'False', 'profiles_dir': '/Users/jeremy/.dbt', 'fail_fast': 'False', 'warn_error': 'None', 'log_path': '/Users/jeremy/src/dbt-basic/logs', 'debug': 'True', 'version_check': 'True', 'use_colors': 'True', 'use_experimental_parser': 'False', 'no_print': 'None', 'quiet': 'False', 'log_format': 'default', 'static_parser': 'True', 'introspect': 'True', 'invocation_command': 'dbt --debug run -s snek', 'target_path': 'None', 'warn_error_options': 'WarnErrorOptions(include=[], exclude=[])', 'send_anonymous_usage_stats': 'True'}
22:03:39  Sending event: {'category': 'dbt', 'action': 'project_id', 'label': '1b3c04de-d16a-426d-a053-f62a87ef3402', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10910dbb0>]}
22:03:39  Sending event: {'category': 'dbt', 'action': 'adapter_info', 'label': '1b3c04de-d16a-426d-a053-f62a87ef3402', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11ed883d0>]}
22:03:39  Registered adapter: snowflake=1.6.2
22:03:39  checksum: 8d66c46a7c5c0e5717f369273dbde9db64740684d9ba42e6442c1808d698ae9f, vars: {}, profile: , target: , version: 1.6.2
22:03:39  Partial parsing not enabled
22:03:40  Sending event: {'category': 'dbt', 'action': 'load_project', 'label': '1b3c04de-d16a-426d-a053-f62a87ef3402', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11ef569a0>]}
22:03:40  Sending event: {'category': 'dbt', 'action': 'resource_counts', 'label': '1b3c04de-d16a-426d-a053-f62a87ef3402', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11ef84160>]}
22:03:40  Found 1 model, 0 sources, 0 exposures, 0 metrics, 373 macros, 0 groups, 0 semantic models
22:03:40  Sending event: {'category': 'dbt', 'action': 'runnable_timing', 'label': '1b3c04de-d16a-426d-a053-f62a87ef3402', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11ef84eb0>]}
22:03:40  
22:03:40  Acquiring new snowflake connection 'master'
22:03:40  Acquiring new snowflake connection 'list_development'
22:03:40  Using snowflake connection "list_development"
22:03:40  On list_development: /* {"app": "dbt", "dbt_version": "1.6.2", "profile_name": "all", "target_name": "sf", "connection_name": "list_development"} */
show terse schemas in database development
    limit 10000
22:03:40  Opening a new connection, currently in state init
22:03:41  SQL status: SUCCESS 281 in 2.0 seconds
22:03:42  On list_development: Close
22:03:42  Re-using an available connection from the pool (formerly list_development, now create_development_dbt_jyeo)
22:03:42  Creating schema "database: "development"
schema: "dbt_jyeo"
"
22:03:42  Using snowflake connection "create_development_dbt_jyeo"
22:03:42  On create_development_dbt_jyeo: /* {"app": "dbt", "dbt_version": "1.6.2", "profile_name": "all", "target_name": "sf", "connection_name": "create_development_dbt_jyeo"} */
create schema if not exists development.dbt_jyeo
22:03:42  Opening a new connection, currently in state closed
22:03:43  SQL status: SUCCESS 1 in 1.0 seconds
22:03:43  On create_development_dbt_jyeo: Close
22:03:43  Re-using an available connection from the pool (formerly create_development_dbt_jyeo, now list_development_dbt_jyeo)
22:03:43  Using snowflake connection "list_development_dbt_jyeo"
22:03:43  On list_development_dbt_jyeo: /* {"app": "dbt", "dbt_version": "1.6.2", "profile_name": "all", "target_name": "sf", "connection_name": "list_development_dbt_jyeo"} */
show terse objects in development.dbt_jyeo limit 10000
22:03:43  Opening a new connection, currently in state closed
22:03:44  SQL status: SUCCESS 0 in 1.0 seconds
22:03:44  On list_development_dbt_jyeo: Close
22:03:45  Sending event: {'category': 'dbt', 'action': 'runnable_timing', 'label': '1b3c04de-d16a-426d-a053-f62a87ef3402', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11ed8c0d0>]}
22:03:45  Concurrency: 1 threads (target='sf')
22:03:45  
22:03:45  Began running node model.my_dbt_project.snek
22:03:45  1 of 1 START python table model dbt_jyeo.snek .................................. [RUN]
22:03:45  Re-using an available connection from the pool (formerly list_development_dbt_jyeo, now model.my_dbt_project.snek)
22:03:45  Began compiling node model.my_dbt_project.snek
22:03:45  Writing injected SQL for node "model.my_dbt_project.snek"
22:03:45  Timing info for model.my_dbt_project.snek (compile): 10:03:45.365803 => 10:03:45.408906
22:03:45  Began executing node model.my_dbt_project.snek
22:03:45  Writing runtime python for node "model.my_dbt_project.snek"
22:03:45  Using snowflake connection "model.my_dbt_project.snek"
22:03:45  On model.my_dbt_project.snek: /* {"app": "dbt", "dbt_version": "1.6.2", "profile_name": "all", "target_name": "sf", "node_id": "model.my_dbt_project.snek"} */
WITH snek__dbt_sp AS PROCEDURE ()

RETURNS STRING
LANGUAGE PYTHON
RUNTIME_VERSION = '3.8'
PACKAGES = ('snowflake-snowpark-python')

HANDLER = 'main'
EXECUTE AS CALLER
AS
$$

import sys
sys._xoptions['snowflake_partner_attribution'].append("dbtLabs_dbtPython")


  
    
# snowflake
import pandas as pd
def model(dbt, session):
    dbt.config(transient=True)
    return pd.DataFrame({"id": [1]})


# This part is user provided model code
# you will need to copy the next section to run the code
# COMMAND ----------
# this part is dbt logic for get ref work, do not modify

def ref(*args, **kwargs):
    refs = {}
    key = '.'.join(args)
    version = kwargs.get("v") or kwargs.get("version")
    if version:
        key += f".v{version}"
    dbt_load_df_function = kwargs.get("dbt_load_df_function")
    return dbt_load_df_function(refs[key])


def source(*args, dbt_load_df_function):
    sources = {}
    key = '.'.join(args)
    return dbt_load_df_function(sources[key])


config_dict = {}


class config:
    def __init__(self, *args, **kwargs):
        pass

    @staticmethod
    def get(key, default=None):
        return config_dict.get(key, default)

class this:
    """dbt.this() or dbt.this.identifier"""
    database = "development"
    schema = "dbt_jyeo"
    identifier = "snek"
    
    def __repr__(self):
        return 'development.dbt_jyeo.snek'


class dbtObj:
    def __init__(self, load_df_function) -> None:
        self.source = lambda *args: source(*args, dbt_load_df_function=load_df_function)
        self.ref = lambda *args, **kwargs: ref(*args, **kwargs, dbt_load_df_function=load_df_function)
        self.config = config
        self.this = this()
        self.is_incremental = False

# COMMAND ----------

# To run this in snowsight, you need to select entry point to be main
# And you may have to modify the return type to text to get the result back
# def main(session):
#     dbt = dbtObj(session.table)
#     df = model(dbt, session)
#     return df.collect()

# to run this in local notebook, you need to create a session following examples https://github.com/Snowflake-Labs/sfguide-getting-started-snowpark-python
# then you can do the following to run model
# dbt = dbtObj(session.table)
# df = model(dbt, session)


def materialize(session, df, target_relation):
    # make sure pandas exists
    import importlib.util
    package_name = 'pandas'
    if importlib.util.find_spec(package_name):
        import pandas
        if isinstance(df, pandas.core.frame.DataFrame):
          session.use_database(target_relation.database)
          session.use_schema(target_relation.schema)
          # session.write_pandas does not have overwrite function
          df = session.createDataFrame(df)
    
    df.write.mode("overwrite").save_as_table('development.dbt_jyeo.snek', create_temp_table=False)

def main(session):
    dbt = dbtObj(session.table)
    df = model(dbt, session)
    materialize(session, df, dbt.this)
    return "OK"

  
$$
CALL snek__dbt_sp();
22:03:45  Opening a new connection, currently in state closed
22:03:50  SQL status: SUCCESS 1 in 5.0 seconds
22:03:50  Timing info for model.my_dbt_project.snek (execute): 10:03:45.410550 => 10:03:50.161700
22:03:50  On model.my_dbt_project.snek: Close
22:03:50  Sending event: {'category': 'dbt', 'action': 'run_model', 'label': '1b3c04de-d16a-426d-a053-f62a87ef3402', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11ff4f130>]}
22:03:50  1 of 1 OK created python table model dbt_jyeo.snek ............................. [SUCCESS 1 in 5.26s]
22:03:50  Finished running node model.my_dbt_project.snek
22:03:50  Connection 'master' was properly closed.
22:03:50  Connection 'model.my_dbt_project.snek' was properly closed.
22:03:50  
22:03:50  Finished running 1 table model in 0 hours 0 minutes and 10.20 seconds (10.20s).
22:03:50  Command end result
22:03:50  
22:03:50  Completed successfully
22:03:50  
22:03:50  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
22:03:50  Command `dbt run` succeeded at 10:03:50.642741 after 12.50 seconds
22:03:50  Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'end', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10a0b9820>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11eed7340>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11eda0250>]}
22:03:50  Flushing usage events
```

3. Check table type in Snowflake:

image

^ Kind is `TABLE` instead of `TRANSIENT`.

### Relevant log output

_No response_

### Environment

```markdown
- OS: macOS
- Python: 3.9.13
- dbt-core: 1.6.2
- dbt-snowflake: 1.6.2
```

### Additional Context

Looks like `save_as_table()` does support `transient` type (https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/latest/api/snowflake.snowpark.DataFrameWriter.save_as_table)

https://github.com/dbt-labs/dbt-snowflake/blob/eb4fd78ce2e4200a7c437c6411b95c30a34b1840/dbt/include/snowflake/macros/materializations/table.sql#L55

I think we just need to get the config and then inject into the python code.

### PR DESCRIPTION
**Pull request by dbeatty10**
resolves #776
branched from #777
No [docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) needed

### Problem

Python models don't respect `transient` config.

1. By default, **SQL** tables are created as [**transient**](https://github.com/dbt-labs/dbt-snowflake/blob/ff08a1cd9ea61bb5e54441a4a84d507ca461eedd/dbt/include/snowflake/macros/relations/table/create.sql#L20-L24) (as long as not `temporary`).
1. By default, **Python** tables are created as [**permanent**](https://github.com/snowflakedb/snowpark-python/blob/653462965ef2db45b085e24b59caa61fc253ae51/src/snowflake/snowpark/dataframe_writer.py#L117-L118) because the `table_type` parameter isn't specified [here](https://github.com/dbt-labs/dbt-snowflake/blob/ff08a1cd9ea61bb5e54441a4a84d507ca461eedd/dbt/include/snowflake/macros/materializations/table.sql#L55).

A side problem is that we are currently using the `create_temp_table` parameter which is deprecated.

### Solution

The changes in this PR will allow Python tables to respect the `transient` config - in line with SQL tables described in (1) above. It also stops using the deprecated `create_temp_table` parameter in favor of using `table_type` instead.

This PR adopts the interpretation that `py_write_table` is meant to be a private helper macro of the table materialization. As such, it's a "naive" helper rather than being stand-alone method.

Under that interpretation, we could probably justify making a backwards-incompatible change to the method signature. But out of an abundance of caution, this PR keeps `py_write_table` as backwards-compatible. The one caveat is that any usage of the old signature will not respect `transient` config.

### Alternative solution

Alternatively, we could just go ahead and make the backwards-incompatible change and declare that `py_write_table` has always been a private macro only intended to be called from within the `table` materialization. It will just create whichever type of table it is passed from within the materialization.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc)

### CODE REVIEW

File: dbt/include/snowflake/macros/relations/table/create.sql [hunk 1]
```diff
@@ -1,6 +1,15 @@
 {% macro snowflake__create_table_as(temporary, relation, compiled_code, language='sql') -%}
+  {%- set transient = config.get('transient', default=true) -%}
+
+  {% if temporary -%}
+    {%- set table_type = "temporary"
```
Comment (mikealfare) [inline dbt/include/snowflake/macros/relations/table/create.sql:5]: All of these `{%- set` need a `%}` at the end

Comment (dbeatty10) [reply dbt/include/snowflake/macros/relations/table/create.sql:5]: thank you ❤️

**Comment by mikealfare** [comment]
@dbeatty10 While this is running, have you thought about what tests would be needed, or how existing testing would cover this change?

**Comment by dbeatty10** [comment]
> @dbeatty10 While this is running, have you thought about what tests would be needed, or how existing testing would cover this change?

I hadn't given the testing any thought.

Just poked around a little bit, and there are a couple tests for python models [here](https://github.com/dbt-labs/dbt-snowflake/blob/ff08a1cd9ea61bb5e54441a4a84d507ca461eedd/tests/functional/adapter/test_python_model.py).

But none of them appeared to verify the underlying table type that Snowflake is using. Tried a quick search for [`transient`](https://github.com/search?q=repo%3Adbt-labs%2Fdbt-snowflake%20transient&type=code) and [`temporary`](https://github.com/search?q=repo%3Adbt-labs%2Fdbt-snowflake+temporary&type=code), and didn't see any other tests in which we are verifying that table type.

So up to you how to approach the testing here. One option would be to outline the cases we care about and do a set of manual tests to verify those cases.

**Review by mikealfare — APPROVED** [review]

**Pull request merged**